crochet-stitcher/src/stages/1-parser.ts

```diff
@@ -1,5 +1,277 @@
+    }
+
+    const incrementLUT = new Set<string>(['increment', 'inc']);
+    if (incrementLUT.has(input)) {
+        return { type: TokenType.Increment, value: input };
+    }
+    const decrementLUT = new Set<string>(['decrement', 'dec']);
+    if (decrementLUT.has(input)) {
+        return { type: TokenType.Decrement, value: input };
+    }
+    const invertLUT = new Set<string>(['invert', 'inv']);
+    if (invertLUT.has(input)) {
+        return { type: TokenType.Invert, value: input };
+    }
+
+    //add more colours here as desired
+    const colourLUT = new Set<string>([
+        'white',
+        'black',
+        'red',
+        'orange',
+        'yellow',
+        'green',
+        'blue',
+        'indigo',
+        'violet',
+    ]);
+    if (colourLUT.has(input)) {
+        return { type: TokenType.Colour, value: input };
+    }
+
+    // Note! if the foundation is specified as a slipknot,
+    // there's no way for us to tell the difference between specifying it
+    // as a stitch vs a foundation knot based on the grammar. As such we'll only catch magic ring
+    const foundationLUT = new Set<string>(['mr', 'magicring', 'magic', 'ring']);
+    if (foundationLUT.has(input)) {
+        return { type: TokenType.Foundation, value: 'mr' };
+    }
+
+    // This should simplify latter stages of the grammar
+    const countAlias = new Set<string>(['twice', 'thrice']);
+    if (countAlias.has(input)) {
+        if (input === 'twice') {
+            return { type: TokenType.Number, value: 2 };
+        }
+        if (input === 'thrice') {
+            return { type: TokenType.Number, value: 3 };
+        }
+    }
+
+    // Misc. aliases
+    const repeat = new Set<string>(['rep', 'repeat']);
+    if (repeat.has(input)) {
+        return { type: TokenType.Repeat, value: 'repeat' };
+    }
+
+    const times = new Set<string>(['time', 'times']);
+    if (times.has(input)) {
+        return { type: TokenType.Times, value: 'times' };
+    }
+
+    // Misc. cases
+    if (input === '*') {
+        return { type: TokenType.Asterisk, value: '*' };
+    }
+    if (input === 'x') {
+        return { type: TokenType.x, value: 'x' };
+    }
+    if (input === 'from') {
+        return { type: TokenType.From, value: 'from' };
+    }
+    if (input === 'more') {
+        return { type: TokenType.More, value: 'more' };
+    }
+
+    // throw new TypeError(`Unknown string token ${input}`)
+    console.warn(`Unknown string token ${input}`);
+    // debug fix CHANGE THIS
+    return { type: TokenType.Colon, value: input };
+}
+
+function tokenizer(input: string): ParsedToken[] {
+    input = input.replace(/\s+/g, ' ').trim();
+    let output: ParsedToken[] = [];
+    let tokens = input.split(' ');
+    for (let i = 0; i < tokens.length; i++) {
+        let val = tokens[i];
+        let endTokens: ParsedToken[] = [];
+        while (val.length > 0) {
+            // strip off any leading brackets (only type of leading thing according to spec)
```

Comment:
Hmm, I wonder if we could just construct a better regex that splits the symbols/numbers automatically.